### PR TITLE
Tokenization refactor

### PR DESCRIPTION
**Pull request by mnunberg**

### CODE REVIEW

File: src/concurrent_ctx.c [hunk 1]
```diff
@@ -67,17 +72,14 @@ void ConcurrentSearchCtx_Init(RedisModuleCtx *rctx, ConcurrentSearchCtx *ctx) {
...
-  clock_gettime(CLOCK_MONOTONIC_RAW, &ctx->lastTime);
+  ConcurrentSearchCtx_ResetClock(rctx, ctx);
 }
 
 void ConcurrentSearchCtx_Free(ConcurrentSearchCtx *ctx) {
   // Release the monitored open keys
   for (size_t i = 0; i < ctx->numOpenKeys; i++) {
-    // Close the monitored key and free its reopen string
-    RedisModule_CloseKey(ctx->openKeys[i].key);
     RedisModule_FreeString(ctx->ctx, ctx->openKeys[i].keyName);
```
Comment (dvirsky) [inline src/concurrent_ctx.c]: Why did you remove closing the key? this might cause a leak

File: src/dep/triemap/triemap.c [hunk 2]
```diff
@@ -37,7 +37,7 @@ TrieMapNode *__trieMapNode_resizeChildren(TrieMapNode *n, int offset) {
 /* Create a new trie node. str is a string to be copied into the node,
  * starting from offset up until len. numChildren is the initial number of
  * allocated child nodes */
-TrieMapNode *__newTrieMapNode(char *str, tm_len_t offset, tm_len_t len, tm_len_t numChildren,
+TrieMapNode *__newTrieMapNode(const char *str, tm_len_t offset, tm_len_t len, tm_len_t numChildren,
```
Comment (dvirsky) [inline src/dep/triemap/triemap.c]: all this is meaningless - this is a git subtree and you didn't change the origin of it, it will just get deleted.

File: src/concurrent_ctx.c [hunk 3]
```diff
@@ -58,7 +56,14 @@ inline void ConcurrentSearch_CheckTimer(ConcurrentSearchCtx *ctx) {
     // This will be used to calculate the elapsed running time
     clock_gettime(CLOCK_MONOTONIC_RAW, &ctx->lastTime);
```
Comment (dvirsky) [inline src/concurrent_ctx.c]: better call ConcurrentSearchCtx_ResetClock

File: src/document.c [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+  return 0;
+
+fail:
+  DocTable_Delete(table, keystr);
+  return -1;
+}
+
+typedef struct {
+  ForwardIndex *idx;
+  RSSortingVector *sv;
```
Comment (mnunberg) [inline src/document.c]: The sorting vector does not become shared until later on, (i.e. until `DocTable_SetSortingVector` is called), and once it's called, it is no longer touched - therefore there is no need in protecting it.

File: src/document.c [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+void Document_FreeDetatched(Document *doc, RedisModuleCtx *anyCtx) {
+  RedisModule_FreeString(anyCtx, doc->docKey);
+
+  for (size_t ii = 0; ii < doc->numFields; ++ii) {
+    RedisModule_FreeString(anyCtx, doc->fields[ii].text);
+    free((void *)doc->fields[ii].name);
+  }
+
+  free((char *)doc->payload);
+  free((char *)doc->language);
```
Comment (mnunberg) [inline src/document.c]: free(NULL) is perfectly acceptable.

File: src/document.c [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+void Document_FreeDetatched(Document *doc, RedisModuleCtx *anyCtx) {
+  RedisModule_FreeString(anyCtx, doc->docKey);
+
+  for (size_t ii = 0; ii < doc->numFields; ++ii) {
+    RedisModule_FreeString(anyCtx, doc->fields[ii].text);
+    free((void *)doc->fields[ii].name);
+  }
+
+  free((char *)doc->payload);
```
Comment (mnunberg) [inline src/document.c]: free(NULL) is perfectly acceptable

File: src/document.c [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+  doc->docKey = docKey;
+  doc->score = (float)score;
+  doc->numFields = numFields;
+  doc->fields = calloc(doc->numFields, sizeof(DocumentField));
+  doc->language = lang;
+  doc->payload = payload;
+  doc->payloadSize = payloadSize;
+}
+
+void Document_Detatch(Document *doc, RedisModuleCtx *srcCtx) {
```
Comment (mnunberg) [inline src/document.c]: Done

File: src/document.c [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+                   const char *lang, const char *payload, size_t payloadSize) {
+  doc->docKey = docKey;
+  doc->score = (float)score;
+  doc->numFields = numFields;
+  doc->fields = calloc(doc->numFields, sizeof(DocumentField));
+  doc->language = lang;
+  doc->payload = payload;
+  doc->payloadSize = payloadSize;
+}
+
```
Comment (mnunberg) [inline src/document.c]: Done

File: src/document.c [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+  if (doc->language) {
+    doc->language = strdup(doc->language);
+  }
+}
+
+void Document_Free(Document *doc) {
+  free(doc->fields);
+}
+
+void Document_FreeDetatched(Document *doc, RedisModuleCtx *anyCtx) {
```
Comment (mnunberg) [inline src/document.c]: Done

File: src/document.c [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+  }
+
+  for (int i = 0; i < doc->numFields; i++) {
+    RedisModule_HashSet(k, REDISMODULE_HASH_CFIELDS, doc->fields[i].name, doc->fields[i].text,
+                        NULL);
+  }
+  return REDISMODULE_OK;
+}
+
+static int Document_Store(Document *doc, RedisSearchCtx *ctx, int options,
```
Comment (mnunberg) [inline src/document.c]: Done

File: src/document.c [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+      RedisModule_ModuleTypeGetType(k) != IndexSpecType) {
+    aCtx->rsCtx.spec = NULL;
+  }
+  aCtx->rsCtx.spec = RedisModule_ModuleTypeGetValue(k);
+}
+
+RSAddDocumentCtx *NewAddDocumentCtx(RedisModuleCtx *origCtx, IndexSpec *sp) {
+  // Block the client and create the context!
+  RSAddDocumentCtx *aCtx = calloc(1, sizeof(*aCtx));
+  aCtx->bc = RedisModule_BlockClient(origCtx, NULL, NULL, NULL, 0);
```
Comment (mnunberg) [inline src/document.c]: I've changed it as requested, but now it looks ugly, because there is more boilerplate. I think blocking the client in AddDocumentCtx is cleaner because there is symmetry when it's unblocked.

File: src/forward_index.h [hunk 12]
```diff
@@ -14,33 +17,34 @@ typedef struct {
...
+  // khash_t(32) * hits;
+  TrieMap *hits;
   t_docId docId;
   uint32_t totalFreq;
   uint32_t maxFreq;
+  uint32_t idxFlags;
   float docScore;
   int uniqueTokens;
   Stemmer *stemmer;
+  BlkAlloc entries;
```
Comment (mnunberg) [inline src/forward_index.h]: Done

File: src/module.c [hunk 13]
```diff
@@ -285,32 +129,35 @@ int AddDocumentCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
   size_t payloadSize = 0;
   RMUtil_ParseArgsAfter("PAYLOAD", argv, argc, "b", &payload, &payloadSize);
 
-  Document doc = NewDocument(argv[2], ds, (argc - fieldsIdx) / 2, lang ? lang : DEFAULT_LANGUAGE,
-                             payload, payloadSize);
+  IndexSpec *sp = IndexSpec_Load(ctx, RedisModule_StringPtrLen(argv[1], NULL), 0);
```
Comment (mnunberg) [inline src/module.c]: The spec is opened for writing later on by ConcurrentSearchCtx

File: src/trie/rune_util.c [hunk 14]
```diff
@@ -51,18 +52,34 @@ rune *strToFoldedRunes(char *str, size_t *len) {
...
+  ret[rlen] = '\0';
+  if (len) {
+    *len = rlen;
   }
-  if (len) *len = rlen;
-
   return ret;
 }
+
+static char *rune_32to16(uint32_t cp, char *encoded) {
```
Comment (mnunberg) [inline src/trie/rune_util.c]: leftover. i'll remove it.

File: src/concurrent_ctx.c (hunk 1 above)
Comment (mnunberg) [inline src/concurrent_ctx.c]: Fixed with adding a flag to check if the key is open or not

File: src/concurrent_ctx.c [hunk 15]
```diff
@@ -1,22 +1,30 @@
...
+      numProcs = CONCURRENT_INDEX_POOL_SIZE;
+    }
+    ConcurrentIndexThreadPool = thpool_init(numProcs);
   }
 }
 
 /* Run a function on the concurrent thread pool */
-void ConcurrentSearch_ThreadPoolRun(void (*func)(void *), void *arg) {
-  thpool_add_work(ConcurrentSearchThreadPool, func, arg);
+void ConcurrentSearch_ThreadPoolRun(void (*func)(void *), void *arg, int type) {
```
Comment (dvirsky) [inline src/concurrent_ctx.c]: I don't like this, I'd rather expose two init functions and two thread pull run functions.

File: src/document.h [hunk 16]
```diff
@@ -19,12 +21,94 @@ typedef struct {
...
+#define ACTX_F_MERGED 0x01
+#define ACTX_F_ERRORED 0x02
 
-void Document_Free(Document doc);
+typedef struct RSAddDocumentCtx {
+  struct RSAddDocumentCtx *next;
+  Document doc;
+  RedisModuleBlockedClient *bc;
+  RedisModuleCtx *thCtx;
+  RedisSearchCtx rsCtx;
```
Comment (dvirsky) [inline src/document.h]: the search context contains a redis module context, maybe thCtx is redundant then?

File: src/forward_index.c [hunk 17]
```diff
@@ -4,116 +4,209 @@
...
+#define CHARS_PER_TERM 5
+static size_t estimtateTermCount(const Document *doc) {
+  size_t nChars = 0;
+  for (size_t ii = 0; ii < doc->numFields; ++ii) {
+    size_t n;
+    RedisModule_StringPtrLen(doc->fields[ii].text, &n);
+    nChars += n;
+  }
+  return nChars / CHARS_PER_TERM;
+  // return 3;
```
Comment (dvirsky) [inline src/forward_index.c]: LOL. remove this please :)

File: src/redis_index.c [hunk 18]
```diff
@@ -95,7 +95,26 @@ int InvertedIndex_RegisterType(RedisModuleCtx *ctx) {
 * TODO: Add index name to it
 */
 RedisModuleString *fmtRedisTermKey(RedisSearchCtx *ctx, const char *term, size_t len) {
-  return RedisModule_CreateStringPrintf(ctx->redisCtx, TERM_KEY_FORMAT, ctx->spec->name, len, term);
+  char buf_s[1024] = {"ft:"};
```
Comment (dvirsky) [inline src/redis_index.c]: I'm okay with limiting the token size. It's actually a good security measure.

File: src/spec.c [hunk 19]
```diff
@@ -239,7 +239,12 @@ IndexSpec *IndexSpec_Parse(const char *name, const char **argv, int argc, char *
 }
 
 int IndexSpec_AddTerm(IndexSpec *sp, const char *term, size_t len) {
-  return Trie_InsertStringBuffer(sp->terms, (char *)term, len, 1, 1, NULL);
+  int isNew = Trie_InsertStringBuffer(sp->terms, (char *)term, len, 1, 1, NULL);
+  if (isNew) {
```
Comment (dvirsky) [inline src/spec.c]: nice touch :)

File: src/tokenize.c [hunk 20]
```diff
@@ -24,13 +24,72 @@ int tokenize(const char *text, float score, t_fieldMask fieldId, void *ctx, Toke
   return _tokenize(&tctx);
 }
 
+static const char sepMap[256] = {
```
Comment (dvirsky) [inline src/tokenize.c]: IIRC you can initilize it with just the 1 members, the 0s are implicit.

File: src/tokenize.c [hunk 21]
```diff
@@ -24,13 +24,72 @@ int tokenize(const char *text, float score, t_fieldMask fieldId, void *ctx, Toke
...
+        [0xD2] = 0, [0xD3] = 0, [0xD4] = 0, [0xD5] = 0, [0xD6] = 0, [0xD7] = 0, [0xD8] = 0,
+        [0xD9] = 0, [0xDA] = 0, [0xDB] = 0, [0xDC] = 0, [0xDD] = 0, [0xDE] = 0, [0xDF] = 0,
+        [0xE0] = 0, [0xE1] = 0, [0xE2] = 0, [0xE3] = 0, [0xE4] = 0, [0xE5] = 0, [0xE6] = 0,
+        [0xE7] = 0, [0xE8] = 0, [0xE9] = 0, [0xEA] = 0, [0xEB] = 0, [0xEC] = 0, [0xED] = 0,
+        [0xEE] = 0, [0xEF] = 0, [0xF0] = 0, [0xF1] = 0, [0xF2] = 0, [0xF3] = 0, [0xF4] = 0,
+        [0xF5] = 0, [0xF6] = 0, [0xF7] = 0, [0xF8] = 0, [0xF9] = 0, [0xFA] = 0, [0xFB] = 0,
+        [0xFC] = 0, [0xFD] = 0, [0xFE] = 0, [0xFF] = 0,
+};
+
+static char *mySep(char **s) {
```
Comment (dvirsky) [inline src/tokenize.c]: make sure this is inlined, it's a very tight loop.

**Comment by dvirsky** [comment]
@mnunberg I'm seeing valgrind errors when indexing: 

```
==14836== Conditional jump or move depends on uninitialised value(s)
==14836==    at 0x7A40A89: nu_utf8_read (utf8.h:37)
==14836==    by 0x7A49666: _nu_strlen (strings.c:12)
==14836==    by 0x7A49666: nu_strlen (strings.c:66)
==14836==    by 0x7A40EE9: strToRunes (rune_util.c:57)
==14836==    by 0x7A42923: runeBufFill (trie_type.c:50)
==14836==    by 0x7A42923: Trie_InsertStringBuffer (trie_type.c:64)
==14836==    by 0x7A3BE24: IndexSpec_AddTerm (spec.c:242)
==14836==    by 0x7A2CDAA: writeCurEntries (document.c:470)
==14836==    by 0x7A2CDAA: DocumentIndexer_Process (document.c:571)
==14836==    by 0x7A2CF67: DocumentIndexer_Run (document.c:616)
==14836==    by 0x5349183: start_thread (pthread_create.c:312)
==14836==    by 0x565CFFC: clone (clone.S:111)
==14836==
==14836== Conditional jump or move depends on uninitialised value(s)
==14836==    at 0x7A49670: _nu_strlen (strings.c:14)
==14836==    by 0x7A49670: nu_strlen (strings.c:66)
==14836==    by 0x7A40EE9: strToRunes (rune_util.c:57)
==14836==    by 0x7A42923: runeBufFill (trie_type.c:50)
==14836==    by 0x7A42923: Trie_InsertStringBuffer (trie_type.c:64)
==14836==    by 0x7A3BE24: IndexSpec_AddTerm (spec.c:242)
==14836==    by 0x7A2CDAA: writeCurEntries (document.c:470)
==14836==    by 0x7A2CDAA: DocumentIndexer_Process (document.c:571)
==14836==    by 0x7A2CF67: DocumentIndexer_Run (document.c:616)
==14836==    by 0x5349183: start_thread (pthread_create.c:312)
==14836==    by 0x565CFFC: clone (clone.S:111)
==14836==
==14836== Conditional jump or move depends on uninitialised value(s)
==14836==    at 0x4C2E0F8: strlen (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14836==    by 0x7A40F04: strToRunes (rune_util.c:59)
==14836==    by 0x7A42923: runeBufFill (trie_type.c:50)
==14836==    by 0x7A42923: Trie_InsertStringBuffer (trie_type.c:64)
==14836==    by 0x7A3BE24: IndexSpec_AddTerm (spec.c:242)
==14836==    by 0x7A2CDAA: writeCurEntries (document.c:470)
==14836==    by 0x7A2CDAA: DocumentIndexer_Process (document.c:571)
==14836==    by 0x7A2CF67: DocumentIndexer_Run (document.c:616)
==14836==    by 0x5349183: start_thread (pthread_create.c:312)
==14836==    by 0x565CFFC: clone (clone.S:111)
==14836==
==14836== Invalid read of size 1
==14836==    at 0x7A40A80: nu_utf8_read (utf8.h:35)
==14836==    by 0x7A49666: _nu_strlen (strings.c:12)
==14836==    by 0x7A49666: nu_strlen (strings.c:66)
==14836==    by 0x7A40EE9: strToRunes (rune_util.c:57)
==14836==    by 0x7A42923: runeBufFill (trie_type.c:50)
==14836==    by 0x7A42923: Trie_InsertStringBuffer (trie_type.c:64)
==14836==    by 0x7A3BE24: IndexSpec_AddTerm (spec.c:242)
==14836==    by 0x7A2CDAA: writeCurEntries (document.c:470)
==14836==    by 0x7A2CDAA: DocumentIndexer_Process (document.c:571)
==14836==    by 0x7A2CF67: DocumentIndexer_Run (document.c:616)
==14836==    by 0x5349183: start_thread (pthread_create.c:312)
==14836==    by 0x565CFFC: clone (clone.S:111)
==14836==  Address 0x496a8780 is 0 bytes after a block of size 144 alloc'd
==14836==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14836==    by 0x7A3F7AE: BlkAlloc_Alloc (block_alloc.c:27)
==14836==    by 0x7A2D713: copyTempString (forward_index.c:109)
==14836==    by 0x7A2D713: forwardIndexTokenFunc (forward_index.c:140)
==14836==    by 0x7A3D1FD: _tokenize (tokenize.c:127)
==14836==    by 0x7A3D2AA: tokenize (tokenize.c:24)
==14836==    by 0x7A2C0EB: fulltextPreprocessor (document.c:178)
==14836==    by 0x7A2C69F: Document_AddToIndexes (document.c:278)
==14836==    by 0x7A4371D: thread_do (thpool.c:337)
==14836==    by 0x5349183: start_thread (pthread_create.c:312)
==14836==    by 0x565CFFC: clone (clone.S:111)
==14836==
==14836== Invalid read of size 1
==14836==    at 0x4C2E0F4: strlen (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14836==    by 0x7A40F04: strToRunes (rune_util.c:59)
==14836==    by 0x7A42923: runeBufFill (trie_type.c:50)
==14836==    by 0x7A42923: Trie_InsertStringBuffer (trie_type.c:64)
==14836==    by 0x7A3BE24: IndexSpec_AddTerm (spec.c:242)
==14836==    by 0x7A2CDAA: writeCurEntries (document.c:470)
==14836==    by 0x7A2CDAA: DocumentIndexer_Process (document.c:571)
==14836==    by 0x7A2CF67: DocumentIndexer_Run (document.c:616)
==14836==    by 0x5349183: start_thread (pthread_create.c:312)
==14836==    by 0x565CFFC: clone (clone.S:111)
==14836==  Address 0x496a8780 is 0 bytes after a block of size 144 alloc'd
==14836==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14836==    by 0x7A3F7AE: BlkAlloc_Alloc (block_alloc.c:27)
==14836==    by 0x7A2D713: copyTempString (forward_index.c:109)
==14836==    by 0x7A2D713: forwardIndexTokenFunc (forward_index.c:140)
==14836==    by 0x7A3D1FD: _tokenize (tokenize.c:127)
==14836==    by 0x7A3D2AA: tokenize (tokenize.c:24)
==14836==    by 0x7A2C0EB: fulltextPreprocessor (document.c:178)
==14836==    by 0x7A2C69F: Document_AddToIndexes (document.c:278)
==14836==    by 0x7A4371D: thread_do (thpool.c:337)
==14836==    by 0x5349183: start_thread (pthread_create.c:312)
==14836==    by 0x565CFFC: clone (clone.S:111)
```

File: src/redis_index.c (hunk 18 above)
Comment (mnunberg) [inline src/redis_index.c]: Me too, but I don't think we should do it in a function ostensibly designed to simply format a string. My other branches have this functionality.
